### PR TITLE
SAMZA-1464: Flushing a closed RocksDB store causes SIGSEGVs

### PR DESCRIPTION
**Pull request by prateekm**
Made RocksDB operations check if DB is still open to avoid segfaults.

### CODE REVIEW

**Comment by prateekm** [comment]
cc @vjagadish1989 @xinyuiscool for review.

File: samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala [hunk 1]
```diff
@@ -164,55 +173,79 @@ class RocksDbKeyValueStore(
...
+      } else {
+        warn(new SamzaException("Close called again on a closed store: %s. Ignoring this close." +
+          "Stack at first close is under 'Caused By'." format storeName, stackAtFirstClose))
+      }
+    } finally {
+      stateChangeLock.writeLock().unlock()
+    }
+  }
+
+  private def withStateCheck[T](fn: => T): T = {
```
Comment (vjagadish1989) [inline samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala]: Minor: cleaner to call it `withCloseGuard` perhaps?

Comment (prateekm) [reply samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala]: How about `ifOpen` or something?

**Comment by prateekm** [comment]
Some numbers from running the new concurrent write test on my MBP with and without these changes. These are not well controlled experiments, but I don't see any obvious perf degradations.

```
With Locks:
4 threads, 1000000 messages
[main] TestKeyValuePerformance$ [INFO] Running test: rocksdb-concurrent-write-performance
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 256 bytes with 4 threads: 4.861 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 512 bytes with 4 threads: 9.721 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 1024 bytes with 4 threads: 12.998000000000001 sec
[main] TestKeyValuePerformance [INFO] Total time to write 4000000 msgs of size 256 bytes with 1 threads: 8.55 sec
[main] TestKeyValuePerformance [INFO] Total time to write 4000000 msgs of size 512 bytes with 1 threads: 15.101 sec
[main] TestKeyValuePerformance [INFO] Total time to write 4000000 msgs of size 1024 bytes with 1 threads: 25.928 sec

8 threads, 8000000 messages
[main] TestKeyValuePerformance$ [INFO] Running test: rocksdb-concurrent-write-performance
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 256 bytes with 8 threads: 22.993000000000002 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 512 bytes with 8 threads: 25.398 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 1024 bytes with 8 threads: 41.051 sec
[main] TestKeyValuePerformance [INFO] Total time to write 8000000 msgs of size 256 bytes with 1 threads: 18.488 sec
[main] TestKeyValuePerformance [INFO] Total time to write 8000000 msgs of size 512 bytes with 1 threads: 31.264 sec
[main] TestKeyValuePerformance [INFO] Total time to write 8000000 msgs of size 1024 bytes with 1 threads: 64.231 sec

32 threads, 32000000 messages
[main] TestKeyValuePerformance$ [INFO] Running test: rocksdb-concurrent-write-performance
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 256 bytes with 32 threads: 119.179 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 512 bytes with 32 threads: 140.784 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 1024 bytes with 32 threads: 186.508 sec
[main] TestKeyValuePerformance [INFO] Total time to write 32000000 msgs of size 256 bytes with 1 threads: 125.158 sec
[main] TestKeyValuePerformance [INFO] Total time to write 32000000 msgs of size 512 bytes with 1 threads: 163.649 sec
[main] TestKeyValuePerformance [INFO] Total time to write 32000000 msgs of size 1024 bytes with 1 threads: 263.729 sec


Without Locks:
4 threads, 1000000 messages
[main] TestKeyValuePerformance$ [INFO] Running test: rocksdb-concurrent-write-performance
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 256 bytes with 4 threads: 5.125 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 512 bytes with 4 threads: 10.417 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 1024 bytes with 4 threads: 13.514000000000001 sec
[main] TestKeyValuePerformance [INFO] Total time to write 4000000 msgs of size 256 bytes with 1 threads: 8.374 sec
[main] TestKeyValuePerformance [INFO] Total time to write 4000000 msgs of size 512 bytes with 1 threads: 17.107 sec
[main] TestKeyValuePerformance [INFO] Total time to write 4000000 msgs of size 1024 bytes with 1 threads: 26.405 sec

8 threads, 8000000 messages
[main] TestKeyValuePerformance$ [INFO] Running test: rocksdb-concurrent-write-performance
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 256 bytes with 8 threads: 24.841 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 512 bytes with 8 threads: 28.748 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 1024 bytes with 8 threads: 39.853 sec
[main] TestKeyValuePerformance [INFO] Total time to write 8000000 msgs of size 256 bytes with 1 threads: 21.443 sec
[main] TestKeyValuePerformance [INFO] Total time to write 8000000 msgs of size 512 bytes with 1 threads: 32.207 sec
[main] TestKeyValuePerformance [INFO] Total time to write 8000000 msgs of size 1024 bytes with 1 threads: 54.419000000000004 sec

32 threads, 32000000 messages
[main] TestKeyValuePerformance$ [INFO] Running test: rocksdb-concurrent-write-performance
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 256 bytes with 32 threads: 108.47500000000001 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 512 bytes with 32 threads: 129.094 sec
[main] TestKeyValuePerformance [INFO] Total time to write 1000000 msgs of size 1024 bytes with 32 threads: 189.90200000000002 sec
[main] TestKeyValuePerformance [INFO] Total time to write 32000000 msgs of size 256 bytes with 1 threads: 134.212 sec
[main] TestKeyValuePerformance [INFO] Total time to write 32000000 msgs of size 512 bytes with 1 threads: 164.27100000000002 sec
[main] TestKeyValuePerformance [INFO] Total time to write 32000000 msgs of size 1024 bytes with 1 threads: 290.944 sec
```

**Comment by prateekm** [comment]
cc @nickpan47 @jmakes FYI.

File: samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala (hunk 1 above)
Comment (vjagadish1989) [reply samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala]: yeah, even better!

File: samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala [hunk 2]
```diff
@@ -103,7 +104,15 @@ class RocksDbKeyValueStore(
...
+    * Reads and writes to this field must be guarded by stateChangeLock.
+    * This is an Exception instead of an Array[StackTraceElement] for ease of logging.
+    */
+  private var stackAtFirstClose: Exception = null
+  private val stateChangeLock = new ReentrantReadWriteLock()
+
+  def get(key: Array[Byte]): Array[Byte] = withStateCheck {
     metrics.gets.inc
     require(key != null, "Null key not allowed.")
     val found = db.get(key)
```
Comment (sborya) [inline samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala]: should we use found.length and not found.size?

**Comment by prateekm** [comment]
@xinyuiscool It's not just closing the store twice, but any operation on the store after a close. I'll sync with you offline to understand the use case you mentioned.

**Comment by prateekm** [comment]
Talked to @xinyuiscool offline. The reader writer lock is for changing the state of the store (i.e. closing it), not for writing (put/flush) to it. get/put/flush/delete all acquire the read lock, so contention shouldn't be a huge issue.

**Pull request closed without merging**